src/Database/LSMTree/Internal/TableHandle.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE DeriveAnyClass #-}
+
+-- TODO: remove once the top-level bindings are in use
+{-# OPTIONS_GHC -Wno-unused-top-binds #-}
+
+module Database.LSMTree.Internal.TableHandle (
+    -- * Exceptions
+    LSMTreeError (..)
+    -- * Session
+  , Session (..)
+    -- * Table handle
+  , TableHandle
+    -- * configuration
+  , TableConfig (..)
+  , ResolveMupsert (..)
+  , SizeRatio (..)
+  , MergePolicy (..)
+  , BloomFilterAlloc (..)
+  ) where
+
+import           Control.Concurrent.Class.MonadMVar.Strict (StrictMVar)
+import           Control.Exception (Exception)
+import           Data.BloomFilter (Bloom)
+import qualified Data.Vector as V
+import           Data.Word (Word32)
+import           Database.LSMTree.Internal.IndexCompact (IndexCompact)
+import           Database.LSMTree.Internal.Run (Run)
+import qualified Database.LSMTree.Internal.Run as Run
+import           Database.LSMTree.Internal.Serialise (SerialisedKey,
+                     SerialisedValue)
+import           Database.LSMTree.Internal.WriteBuffer (WriteBuffer)
+import           System.FS.API (FsErrorPath, FsPath, HasFS)
+import           System.FS.BlockIO.API (HasBlockIO)
+
+{-------------------------------------------------------------------------------
+  Exceptions
+-------------------------------------------------------------------------------}
+
+-- TODO: give this a nicer Show instance.
+data LSMTreeError =
+    SessionDirDoesNotExist FsErrorPath
+  | SessionDirLocked FsErrorPath
+  deriving (Show, Exception)
+
+{-------------------------------------------------------------------------------
+  Session
+-------------------------------------------------------------------------------}
+
+-- | A session provides context that is shared across multiple table handles.
+--
+-- For more information, see 'Database.LSMTree.Common.Session'.
+data Session m h = Session {
+    -- | The path to the directory in which this sesion is live. This is a path
+    -- relative to root of the 'HasFS' instance.
+    sessionRoot        :: !FsPath
+  , sessionHasFS       :: !(HasFS m h)
+  , sessionHasBlockIO  :: !(HasBlockIO m h)
+    -- | A session-wide shared, atomic counter that is used to produce unique
+    -- names for run files.
+  , sessionUniqCounter :: !(StrictMVar m Integer)
+  }
+
+{-------------------------------------------------------------------------------
+  Table handle
+-------------------------------------------------------------------------------}
+
+-- | A handle to an on-disk key\/value table.
+--
+-- For more information, see 'Database.LSMTree.Normal.TableHandle'.
+--
+-- NOTE: a table handle is __not__ thread-safe.
```

Comment:
We might want to be clearer about the guarantees we want to have in the future. If this was not considered thread-safe at all, we could just go with `IORef`s, but it seems like we do in fact ensure some synchronisation.

Reply:
I'm using `MVar`s here because we don't have `IORef`s (yet) in `io-sim`.

Let's iterate on the thread safety guarantees in later PRs